File: Tests/Categories/TestUIAlertView.m
#import <PromiseKit/PromiseKit.h>
#import "UIAlertView+AnyPromise.h"
@import UIKit;
@import XCTest;

@implementation Test_UIAlertView_ObjC: XCTestCase

// fulfills with buttonIndex
- (void)test1 {
    id ex = [self expectationWithDescription:@""];

    UIAlertView *alert = [UIAlertView new];
    [alert addButtonWithTitle:@"0"];
    [alert addButtonWithTitle:@"1"];
    alert.cancelButtonIndex = [alert addButtonWithTitle:@"2"];
    [alert promise].then(^(id obj){
        XCTAssertEqual([obj integerValue], 1);
        [ex fulfill];
    });
    PMKAfter(0.5).then(^{
        [alert dismissWithClickedButtonIndex:1 animated: NO];
    });
    [self waitForExpectationsWithTimeout:3 handler: nil];
}

// cancel button presses are cancelled errors
- (void)test2 {
    id ex = [self expectationWithDescription:@""];

    UIAlertView *alert = [UIAlertView new];
    [alert addButtonWithTitle:@"0"];
    [alert addButtonWithTitle:@"1"];
    alert.cancelButtonIndex = [alert addButtonWithTitle:@"2"];
    [alert promise].catchWithPolicy(PMKCatchPolicyAllErrors, ^(NSError *err){
        XCTAssertTrue(err.cancelled);
        [ex fulfill];
    });
    PMKAfter(0.5).then(^{
        [alert dismissWithClickedButtonIndex:2 animated: NO];
    });
    [self waitForExpectationsWithTimeout:3 handler: nil];
}

// single button UIAlertViews don't get considered cancelled
- (void)test3 {
    id ex = [self expectationWithDescription:@""];

    UIAlertView *alert = [UIAlertView new];
    [alert addButtonWithTitle:@"0"];
    [alert promise].then(^{
        [ex fulfill];
    });
    PMKAfter(0.5).then(^{
        [alert dismissWithClickedButtonIndex:0 animated: NO];
    });
    [self waitForExpectationsWithTimeout:3 handler: nil];
}

// single button UIAlertViews don't get considered cancelled unless the cancelIndex is set
- (void)test4 {
    id ex = [self expectationWithDescription:@""];

    UIAlertView *alert = [UIAlertView new];
    alert.cancelButtonIndex = [alert addButtonWithTitle:@"0"];
    [alert promise].catchWithPolicy(PMKCatchPolicyAllErrors, ^(NSError *err){
        [ex fulfill];
    });
    PMKAfter(0.5).then(^{
        [alert dismissWithClickedButtonIndex:0 animated: NO];
    });
    [self waitForExpectationsWithTimeout:3 handler: nil];
}

@end

File: Tests/Categories/TestUIAlertView.swift
import PromiseKit
import UIKit
import XCTest

class Test_UIAlertView_Swift: UIKitTestCase {
    // fulfills with buttonIndex
    func test1() {
        let ex = expectation(description: "")

        let alert = UIAlertView()
        alert.addButton(withTitle: "0")
        alert.addButton(withTitle: "1")
        alert.cancelButtonIndex = alert.addButton(withTitle: "2")
        alert.promise().then { x -> Void in
            XCTAssertEqual(x, 1)
            ex.fulfill()
        }
        after(interval: 0.1).then {
            alert.dismiss(withClickedButtonIndex: 1, animated: false)
        }
        waitForExpectations(timeout: 3, handler: nil)
    }

    // cancel button presses are cancelled errors
    func test2() {
        let ex = expectation(description: "")

        let alert = UIAlertView()
        alert.addButton(withTitle: "0")
        alert.addButton(withTitle: "1")
        alert.cancelButtonIndex = alert.addButton(withTitle: "2")
        alert.promise().catch(policy: .allErrors) { err in
            guard let err = err as? UIAlertView.Error else { return XCTFail() }
            XCTAssertTrue(err == .cancelled)
            XCTAssertTrue(err.isCancelled)
            ex.fulfill()
        }
        after(interval: 0.1).then {
            alert.dismiss(withClickedButtonIndex: 2, animated: false)
        }
        waitForExpectations(timeout: 3, handler: nil)
    }

    // single button UIAlertViews don't get considered cancelled
    func test3() {
        let ex = expectation(description: "")

        let alert = UIAlertView()
        alert.addButton(withTitle: "0")
        alert.promise().then { _ in
            ex.fulfill()
        }
        after(interval: 0.1).then {
            alert.dismiss(withClickedButtonIndex: 0, animated: false)
        }
        waitForExpectations(timeout: 3, handler: nil)
    }

    // single button UIAlertViews don't get considered cancelled unless the cancelIndex is set
    func test4() {
        let ex = expectation(description: "")

        let alert = UIAlertView()
        alert.cancelButtonIndex = alert.addButton(withTitle: "0")
        alert.promise().catch(policy: .allErrors) { _ in
            ex.fulfill()
        }
        after(interval: 0.1).then {
            alert.dismiss(withClickedButtonIndex: 0, animated: false)
        }
        waitForExpectations(timeout: 3, handler: nil)
    }
}
